app/main/views/g_cloud.py

```diff
@@ -544,6 +544,59 @@ def end_search(framework_framework, project_id):
     )
 
 
+@direct_award.route(
+    '/<string:framework_framework>/projects/<int:project_id>/did-you-award-contract',
+    methods=['GET', 'POST']
+)
+def did_you_award_contract(framework_framework, project_id):
+    all_frameworks = data_api_client.find_frameworks().get('frameworks')
+    framework = framework_helpers.get_latest_live_framework(all_frameworks, framework_framework)
+
+    # Get the requested Direct Award Project.
+    project = data_api_client.get_direct_award_project(project_id=project_id)['project']
+
```

Comment:
This needs a check to not show this page before the search has been locked.

```
project = data_api_client.get_direct_award_project(project_id=project_id)['project']
    if not is_direct_award_project_accessible(project, current_user.id):
        abort(404)

    if not project['lockedAt']:
        abort(400)
```

Reply:
It may also need a test to go with this

Reply:
Please have a look at https://github.com/alphagov/digitalmarketplace-buyer-frontend/pull/763/files#diff-74c43b483b6b57229728b027e18f534eR502 for reference

Reply:
It's possible to over-engineer this: if there's no UI for something, and @risicle is also implementing a test on the back-end POST to check that, then having a 400 isn't massively valuable.

The argument goes something like this:

 - we need to check this in the back end anyway;
 - we should assume that users _aren't_ trying to hack our URL's around, and that if this somehow happens then it's _our_ mistake in wiring up the front-end, not the user's mistake;
 - if we implement the check in the back end, we will therefore get 500's from the front-end if the back-end generates an unexpected 4xx
 - we will then get to find out if we did something wrong in the front-end, rather than just keep telling the user that they did something wrong but not really _what_ (when _we should assume_ it's actually _our_ problem).

Reply:
So I just did some tests on local development, and there seems to be a discrepancy.

If I am logged in as a buyer, and try to access `/buyers/direct-award/g-cloud/projects/314159/did-you-award-contract`, then I get a 404, because 314159 is not a valid project id.

When I run a test that accesses `/buyers/direct-award/g-cloud/projects/314159/did-you-award-contract`, I get a 200, even though 314159 is not a valid project id.

I think the problem here is with the patch on `data_api_client.get_direct_award_project`; it returns even if the project id is invalid, whereas the real method raises a 404 if the project id does not exist.

Obviously the patch can be changed to 404 for an individual test, but this might lead to believe an additional check is needed in the implementation, rather than realising that an extra step is needed in the test.

tl;dr I think we need to change the test so that `data_api_client.get_direct_award_project` so that it checks the `project_id` argument

Reply:
@galund  - The reason why I would argue otherwise regarding blocking the user from accessing the screen without being on that stage is because of the following:
- For consistency - this check is done on the other functions (https://github.com/alphagov/digitalmarketplace-buyer-frontend/pull/761/files#diff-c97e3489337a6c42520133356a12e560R604)
- User's should not be able to view a page which they cannot do anything with. I found that user behaviour has dramatically changed over the years and users will try to take a given url and change the ID to try and bypass stages.
- If for any reason the backend checks ever do fail, we have an additional check.

I'm happy to approve if you still think it's fine.

Reply:
Sure - consistency argument probably wins out, would be interested to hear what @risicle thinks tho.

There's definitely something amiss with an architecture that requires every piece of business logic to be checked in two or even three places tho :-)

Reply:
I've added tests and the code @DilwoarH suggested that address this comment

Reply:
Hey @lfdebrux - it seems like this check is missing:

`if not is_direct_award_project_accessible(project, current_user.id):
        abort(404)`

(not sure why we do this, it should be done on the project call level)

Reply:
I think this goes back to what @galund was saying.

I'll add the logic for consistency, but it does seem a bit odd; especially because the only way I can think of testing this is to deliberately have the api call return bad data.